Deprecate Unicode#downcase/upcase/swapcase.

frodsan opened the pull request:
- Deprecates `ActiveSupport::Multibyte::Unicode#downcase/upcase/swapcase` in favor of using String methods directly.
- Removes `ActiveSupport::Multibyte::Chars` wrappers around the following String methods: `#downcase`, `#upcase`, `#swapcase`, and `#capitalize`.

frodsan committed:
Deprecate Unicode#downcase/upcase/swapcase.

Use String methods directly instead.

jeremy commented:
Related to #26743 and #26403.

When you already have a Chars object, this deprecation message tells you to use a method from a different class. How will a user understand this and what should they do?

Is the underlying intent to deprecate ActiveSupport::Multibyte::Chars itself?

frodsan commented:
The intent is to deprecate the wrappers for the String methods (upcase, downcase, swapcase) in AS::Multibyte::Unicode. These methods are used in AS::Multibyte::Chars so I changed to those to just use the String methods directly. The warning only appears if you use the Unicode methods directly. I added a test case to show this behavior, but here is the output in irb:

```
>> ActiveSupport::Multibyte::Unicode.downcase("DEPRECATED")
deprecated
DEPRECATION WARNING: ActiveSupport::Multibyte::Unicode#downcase is deprecated and will be removed from Rails 6.1. Use String#downcase instead. (called from block (3 levels) in run at /Users/frodsan/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/minitest-5.11.3/lib/minitest/test.rb:98)
>> "NOTDEPRECATED".mb_chars.downcase.to_s
notdeprecated
```

I think these could be the wrapper methods referred here: https://github.com/rails/rails/pull/26743#issuecomment-366869537

jeremy commented:
That sounds good. Let's deprecate these methods together as a group and change the warning message to suggest using String directly since the ActiveSupport::Multibyte::Unicode wrapper is no longer needed.

jeremy commented:
/cc @mtsmfm @amatsuda

mtsmfm commented:
:+1:

frodsan commented:
@jeremy I made the changes. I hope I got it right :) Also, I removed the method definitions from `Multibyte::Chars` because that logic is already being handled by `#method_missing`.

frodsan committed:
Use native String#capitalize

frodsan commented:
I'm also removing the wrapper around String#capitalize after discussing this with @mtsmfm (https://github.com/rails/rails/pull/26743#issuecomment-428789819)

mtsmfm commented:
`reverse` can be `chars(@wrapped_string.scan(/\X/).reverse.join)`

jeremy reviewed (approved):
Thank you @frodsan. For those interested, let's look at deprecating `Unicode.normalize` and changing `Chars#reverse` implementation in new PRs.

jeremy commented:
Looks like a [failing test remains](https://travis-ci.org/rails/rails/jobs/440041853#L1307):
```
MultibyteCharsUTF8BehaviourTest#test_method_works_for_proxyed_methods:
FrozenError: can't modify frozen String
    /home/travis/build/rails/rails/activesupport/lib/active_support/multibyte/chars.rb:61:in `capitalize!'
    /home/travis/build/rails/rails/activesupport/lib/active_support/multibyte/chars.rb:61:in `method_missing'
    /home/travis/build/rails/rails/activesupport/test/multibyte_chars_test.rb:481:in `call'
    /home/travis/build/rails/rails/activesupport/test/multibyte_chars_test.rb:481:in `test_method_works_for_proxyed_methods'
```

frodsan commented:
@jeremy I fixed the failing test. 

I just found that the bang version of these methods will raise FrozenError when the wrapped string is frozen. I think that should be expected:

```
>> (-"hello").mb_chars.chomp!
FrozenError (can't modify frozen String)
>> (-"hello").mb_chars.gsub!("m", "")
FrozenError (can't modify frozen String)
>> (-"hello").mb_chars.downcase!
FrozenError (can't modify frozen String)
```
